go/config/config.go

```diff
@@ -105,39 +124,148 @@ func NewConfiguration() *Configuration {
+		if !sec.HasKey("port") {
+			log.Fatal("Cannot read port from config file:", fileName, err)
+		}
+		if !sec.HasKey("datadir") {
+			log.Fatal("Cannot read datadir from config file:", fileName)
+		}
+		if sec.Haskey("log-error") {
+			Config.MySQLErrorLog = sec.Key("log-error").String()
+		} else {
+			Config.MySQLErrorLog = sec.Key("log_error").String()
+		}
+		Config.MySQLPort, _ = sec.Key("port").Int()
+		Config.MySQLDataDir = sec.Key("datadir").String()
+		Config.MySQLInnoDBLogDir = sec.Key("innodb_log_group_home_dir").String()
+		if _, ok := confFiles["MySQL"]; !ok {
+			confFiles["MySQL"] = fileName
+		}
+		log.Infof("Read config: %s", fileName)
+	}
+	return Config, err
+}
+
+// AddKeyToMySQLConfig add new key with value to my.cnf
+func AddKeyToMySQLConfig(key string, value string) error {
+	cfg, err := ini.LoadSources(ini.LoadOptions{AllowBooleanKeys: true, AllowShadows: true}, confFiles["MySQL"])
+	if err == nil {
+		sec := cfg.Section("mysqld")
+		if sec.Haskey(key) {
+			existingValues := sec.Key(key).ValueWithShadows()
+			for _, existingValue := range existingValues {
+				if value == existingValue {
+					return nil
+				}
+			}
+		}
+		_, err := sec.NewKey(key, value)
+		if err != nil {
+			return log.Errore(err)
+		}
+
+		cfg.SaveTo(confFiles["MySQL"])
+	}
+	return err
+}
+
 // Read reads configuration from zero, either, some or all given files, in order of input.
 // A file can override configuration provided in previous file.
-func Read(file_names ...string) *Configuration {
-	for _, file_name := range file_names {
-		read(file_name)
+func Read(fileNames []string, configtype string) *Configuration {
+	if configtype == "appconfig" {
+		for _, fileName := range fileNames {
+			readJSON(fileName)
+		}
+	}
+	if configtype == "mysqlconfig" {
+		for _, fileName := range fileNames {
+			readINI(fileName)
+		}
 	}
 	return Config
 }
 
 // ForceRead reads configuration from given file name or bails out if it fails
-func ForceRead(file_name string) *Configuration {
-	_, err := read(file_name)
+func ForceRead(fileName string) *Configuration {
+	_, err := readJSON(fileName)
 	if err != nil {
-		log.Fatal("Cannot read config file:", file_name, err)
+		log.Fatal("Cannot read config file:", fileName, err)
 	}
 	return Config
 }
+
+// WatchConf watches for changes in configuration files and rereads them in case of change
+func WatchConf() {
+	watcher, err := fsnotify.NewWatcher()
+	if err == nil {
+		defer watcher.Close()
+		done := make(chan bool)
+		go func() {
+			for {
+				select {
+				case event := <-watcher.Events:
+					if event.Op&fsnotify.Write == fsnotify.Write {
+						if filepath.Base(event.Name) == "my.cnf" {
+							log.Infof("MySQL config file %s changed. Reloading", event.Name)
+							readINI(event.Name)
+						}
+						if filepath.Base(event.Name) == "orchestrator-agent.conf.json" {
+							log.Infof("Orchestrator agent config file %s changed. Reloading", event.Name)
+							readJSON(event.Name)
```

Comment:
Thanks for pointing on this. I've checked the code one more time, and it seems like reloading configs for orchestrator-agent isn't quite a good idea, because some config values (like HTTPPort) can't be change in runtime, so it will be quite confusing to have only part of config values to be reloadable on the fly. 
But as for my.cnf i think it's good to have this option, because it's external config file for orchestrator-agent and it can be quite unclear for user that he had to restart orchestrator-agent in case of my.cnf changes. So i think maybe it will be a good idea to add sync.Mutex to Configuration struct and lock and unlock it for my.cnf config values read\writes?

Reply:
Added RWMutex in f4eefa0 in order to protect f4eefa0 struct. Will add RLock\RUnlock too all reads of Config.MySQLErrorLog, Config.MySQLPort, Config.MySQLDataDir, Config.MySQLInnoDBLogDir in other packages